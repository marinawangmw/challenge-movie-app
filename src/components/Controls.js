import React from 'react';
import { View, Text, StyleSheet, Image, TouchableOpacity } from 'react-native';
import { useTheme } from '@react-navigation/native';
import { spacing, typography } from '@/theme';
import { strings } from '@/localization';

export const Controls = ({ controlDatas, controlStyles }) => {
  const { colors } = useTheme();
  const styles = { ...defaultStyles, ...controlStyles };

  const renderControls = (control, idx) => (
    <View style={styles.controlWrapper} key={idx}>
      <TouchableOpacity onPress={control.handleControlPress}>
        <Image
          source={control.icon}
          style={[styles.icon, control.label === strings.controls.play && styles.playIcon]}
          resizeMode="cover"
          accessibilityIgnoresInvertColors
        />
      </TouchableOpacity>
      <Text style={[{ color: colors.text }, typography.label]}>{control.label}</Text>
    </View>
  );

  return (
    <View style={[styles.container, { color: colors.text }]}>
      {controlDatas.map((control, idx) => renderControls(control, idx))}
    </View>
  );
};

export const defaultStyles = StyleSheet.create({
  icon: {
    width: 28,
    height: 28,
    margin: spacing.xs,
  },
  playIcon: {
    height: 34,
    marginTop: spacing.xs,
    marginBottom: 2,
  },
});
