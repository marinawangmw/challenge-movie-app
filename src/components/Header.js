import React from 'react';
import { View, StyleSheet, Image } from 'react-native';
import { spacing } from '@/theme';

export const Header = () => {
  return (
    <View style={styles.header}>
      <Image
        source={require('@/assets/logo/logo.png')}
        style={styles.headerImg}
        resizeMode="contain"
        accessibilityIgnoresInvertColors
      />
    </View>
  );
};

const styles = StyleSheet.create({
  header: {
    marginTop: spacing.xl,
    alignItems: 'center',
    justifyContent: 'center',
  },
  headerImg: {
    height: 40,
    width: 76,
    marginTop: spacing.s,
  },
});
