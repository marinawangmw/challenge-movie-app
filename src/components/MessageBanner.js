import React from 'react';
import { View, Text } from 'react-native';
import { useTheme } from '@react-navigation/native';

export const MessageBanner = ({
  message = "There's no movie added to My List yet",
  customStyles,
}) => {
  const { colors } = useTheme();

  return (
    <View style={[customStyles.messageContainer]}>
      <Text style={{ color: colors.text }}>{message}</Text>
    </View>
  );
};
