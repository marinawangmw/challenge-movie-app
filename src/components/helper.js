import { Linking } from 'react-native';
import { Config } from 'react-native-config';
import Toast from 'react-native-toast-message';
import { getMovieTrailer } from '@/controllers/MovieController';

export const playTrailer = async (id) => {
  try {
    const trailerData = await getMovieTrailer(id);

    if (trailerData.results.length) {
      const url = Config.VIDEO_EXTERNAL_BASE_URL + trailerData.results[0].key;

      const supported = await Linking.canOpenURL(url);

      if (supported) {
        await Linking.openURL(url);
      } else {
        Toast.show({
          type: 'error',
          text1: "Something's wrong",
          text2: `Don't know how to open this URL: ${url}`,
        });
      }
    } else {
      Toast.show({
        type: 'error',
        text1: 'Oops',
        text2: "There's no trailer found for this movie",
      });
    }
  } catch (e) {
    console.log(e);
  }
};
