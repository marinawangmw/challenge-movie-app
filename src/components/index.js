export { Button } from '@/components/Button';
export { ErrorView } from '@/components/ErrorView';
export { TabBarIcon } from '@/components/TabBarIcon';
export { TextField } from '@/components/TextField';
export { Hero } from '@/components/Hero';
export { Header } from '@/components/Header';
export { MovieList } from '@/components/MovieList';
