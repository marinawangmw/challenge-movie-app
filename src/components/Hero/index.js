import React from 'react';
import { View, ImageBackground, StyleSheet } from 'react-native';
import { useSelector, useDispatch } from 'react-redux';
import { useTheme } from '@react-navigation/native';
import { Config } from 'react-native-config';
import LinearGradient from 'react-native-linear-gradient';
import Labels from './Labels';
import { Header, SpecialBanner, Controls } from '@/components';
import { spacing, typography } from '@/theme';
import { getHeroPoster } from '@/selectors/MovieListSelectors';
import { addIcon, playIcon, infoIcon } from '@/assets';
import { strings } from '@/localization';
import { addToMyList } from '@/actions/MovieListActions';
import { NAVIGATION } from '@/constants';

export const Hero = ({ navigation }) => {
  const { colors } = useTheme();
  const heroPoster = useSelector(getHeroPoster);
  const dispatch = useDispatch();

  const handleMyListIconPress = () => {
    dispatch(addToMyList(heroPoster.posterMovie));
  };

  const handleInfoIconPress = () => {
    navigation.navigate(NAVIGATION.movieDetail, {
      movieDetails: heroPoster.posterMovie,
    });
  };

  const controlDatas = [
    { icon: addIcon, label: strings.controls.myList, handleControlPress: handleMyListIconPress },
    { icon: playIcon, label: strings.controls.play },
    { icon: infoIcon, label: strings.controls.info, handleControlPress: handleInfoIconPress },
  ];

  if (heroPoster && heroPoster.posterMovie) {
    return (
      <View style={[styles.container, { backgroundColor: colors.primary }]}>
        <ImageBackground
          source={{ uri: Config.IMAGE_API_BASE_URL + heroPoster.posterMovie.poster_path }}
          resizeMode="cover"
          style={styles.imgBg}
        >
          <LinearGradient
            locations={[0.0001, 1]}
            colors={['rgba(0, 0, 0, 0.0001)', '#000']}
            style={styles.gradient}
          >
            <Header />
            <View style={styles.bottomContainer}>
              <Labels labels={heroPoster.posterGenres} />
              <SpecialBanner
                styles={{
                  banner: { ...styles.banner, backgroundColor: `${colors.secondary}4A` },
                  bannerLabel: [styles.bannerLabel, typography.label, { color: colors.secondary }],
                }}
              />
              <Controls controlDatas={controlDatas} controlStyles={controlStyles} />
            </View>
          </LinearGradient>
        </ImageBackground>
      </View>
    );
  }

  return null;
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  gradient: {
    width: '100%',
    height: '100%',
  },
  imgBg: {
    width: '100%',
  },
  bottomContainer: {
    alignItems: 'center',
    alignSelf: 'center',
    position: 'absolute',
    bottom: 0,
    padding: spacing.s,
  },
  banner: {
    marginVertical: spacing.s,
    borderRadius: 5,
    paddingVertical: 4,
    paddingHorizontal: 60,
  },
  bannerLabel: {
    textTransform: 'uppercase',
    letterSpacing: 0.5,
  },
});

const controlStyles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    width: '100%',
    justifyContent: 'space-around',
    paddingHorizontal: spacing.xl,
    marginVertical: spacing.xs,
  },
  controlWrapper: {
    padding: spacing.xs,
    alignItems: 'center',
    justifyContent: 'center',
    flex: 1,
  },
});
