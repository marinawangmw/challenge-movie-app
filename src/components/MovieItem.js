import React from 'react';
import { StyleSheet, ImageBackground, Image, TouchableOpacity } from 'react-native';
import { Config } from 'react-native-config';
import { useTheme } from '@react-navigation/native';
import { spacing } from '@/theme';
import { movie } from '@/assets';
import { SpecialBanner } from '@/components';
import { NAVIGATION } from '@/constants';

export const MovieItem = ({ item, customStyles, navigation }) => {
  const { colors } = useTheme();
  const styles = { ...defaultStyles, ...customStyles };

  const handleNavigate = () => {
    navigation.navigate(NAVIGATION.movieDetail, {
      movieDetails: item.item,
    });
  };

  return (
    <>
      {item && item.item.poster_path ? (
        <TouchableOpacity onPress={handleNavigate} style={styles.poster}>
          <ImageBackground
            source={{ uri: Config.IMAGE_API_BASE_URL + item.item.poster_path }}
            resizeMode="cover"
            style={styles.img}
          >
            <SpecialBanner
              styles={{
                banner: { ...styles.banner, backgroundColor: colors.secondary },
                bannerLabel: [styles.bannerLabel, { color: colors.text }],
              }}
            />
          </ImageBackground>
        </TouchableOpacity>
      ) : (
        <TouchableOpacity
          style={[
            styles.img,
            styles.unknownContainer,
            { backgroundColor: colors.posterBackground },
          ]}
        >
          <Image
            source={movie}
            resizeMode="cover"
            style={styles.unknownPosterIcon}
            accessibilityIgnoresInvertColors
          />
        </TouchableOpacity>
      )}
    </>
  );
};

const defaultStyles = StyleSheet.create({
  title: {
    marginVertical: spacing.s,
  },
  img: {
    width: 90,
    height: 150,
    margin: 1,
  },
  unknownContainer: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  unknownPosterIcon: {
    width: 50,
    height: 50,
  },
});
