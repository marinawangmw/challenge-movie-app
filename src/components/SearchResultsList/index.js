import React from 'react';
import { StyleSheet, Text, FlatList } from 'react-native';
import { useSelector } from 'react-redux';
import { useTheme } from '@react-navigation/native';
import SearchResultItem from './SearchResultItem';
import { Loading } from '@/screens';
import { MessageBanner } from '@/components';
import { typography, spacing } from '@/theme';
import { en } from '@/localization/en';
import { TYPES } from '@/actions/MovieListActions';
import { isLoadingSelector } from '@/selectors/StatusSelectors';

export const SearchResultsList = ({ data, navigation, title }) => {
  const { colors } = useTheme();
  const isLoading = useSelector((state) => isLoadingSelector([TYPES.FETCH_MOVIE_LISTS], state));

  const renderEmptyMessage = () => {
    return <MessageBanner message={en.search.noObjectMessage} customStyles={message} />;
  };

  const renderHeader = () => {
    if (isLoading) {
      return <Loading />;
    }

    if (title) {
      return <Text style={[{ color: colors.text }, styles.title, typography.title]}>{title}</Text>;
    }

    return null;
  };

  return (
    <FlatList
      data={data}
      renderItem={(item) => <SearchResultItem navigation={navigation} item={item} />}
      keyExtractor={(_item, idx) => idx}
      contentContainerStyle={styles.container}
      ListHeaderComponent={renderHeader}
      ListEmptyComponent={renderEmptyMessage}
    />
  );
};

export const styles = StyleSheet.create({
  container: {
    flexGrow: 1,
    padding: spacing.s,
  },
  title: {
    paddingVertical: spacing.s,
  },
});

export const message = StyleSheet.create({
  messageContainer: {
    flexGrow: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
