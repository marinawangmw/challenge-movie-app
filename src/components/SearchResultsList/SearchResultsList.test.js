import React from 'react';
import { render } from '@testing-library/react-native';
import { SearchResultsList } from '@/components';
import { withProviders } from '@/test-utils';
import { movieList, navigation } from '@/test-utils/helper';

describe('Search Results List renders correcly', () => {
  it('with empty data and title', () => {
    const title = '';
    const data = [];
    const { toJSON } = render(
      withProviders(<SearchResultsList title={title} data={data} navigation={navigation} />)
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('with data and title with data', () => {
    const title = 'test';
    const data = [movieList.movieList];

    const { toJSON } = render(
      withProviders(<SearchResultsList title={title} data={data} navigation={navigation} />)
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
