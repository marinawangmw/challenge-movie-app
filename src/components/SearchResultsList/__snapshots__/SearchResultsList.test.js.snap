// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Results List renders correcly with data and title with data 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  ListHeaderComponent={[Function]}
  contentContainerStyle={
    Object {
      "flexGrow": 1,
      "padding": 16,
    }
  }
  data={
    Array [
      Array [
        Object {
          "adult": false,
          "backdrop_path": "/ziTxBaF1AhbUwcv2IBiFJAf54qD.jpg",
          "id": 65086,
          "overview": "The story follows a young lawyer, Arthur Kipps.",
          "popularity": 16.334,
          "poster_path": "/aJbeDPUwDe9MLpseTFrNTBaQPzf.jpg",
          "release_date": "2012-02-03",
          "title": "The Woman in Black",
          "vote_average": 6.11,
        },
        Object {
          "adult": false,
          "id": 102899,
          "overview": "Armed with the astonishing ability to shrink in scale but increase in strength.",
          "popularity": 69.704,
          "poster_path": "/rS97hUJ1otKTTripGwQ0ujbuIri.jpg",
          "release_date": "2015-07-14",
          "title": "Ant-Man",
          "vote_average": 7.088,
        },
      ],
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <Text
        style={
          Array [
            Object {
              "color": "rgb(28, 28, 30)",
            },
            Object {
              "paddingVertical": 16,
            },
            Object {
              "fontSize": 18,
              "fontWeight": "700",
            },
          ]
        }
      >
        test
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(255, 255, 255)",
            "flexDirection": "row",
            "margin": 2,
            "opacity": 1,
          }
        }
      >
        <Image
          accessibilityIgnoresInvertColors={true}
          resizeMode="cover"
          source={
            Object {
              "uri": NaN,
            }
          }
          style={
            Object {
              "height": 95,
              "width": 160,
            }
          }
        />
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "column",
              "flexWrap": "wrap",
              "padding": 16,
            }
          }
        >
          <View
            style={
              Object {
                "paddingBottom": 16,
                "width": "100%",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontSize": 16,
                    "fontWeight": "400",
                  },
                  Object {
                    "color": "rgb(28, 28, 30)",
                  },
                  Object {
                    "flexWrap": "wrap",
                  },
                ]
              }
            />
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            />
            <Image
              accessibilityIgnoresInvertColors={true}
              resizeMode="contain"
              source={
                Object {
                  "testUri": "../../../src/assets/play.png",
                }
              }
              style={
                Object {
                  "height": 20,
                  "width": 20,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Search Results List renders correcly with empty data and title 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  ListHeaderComponent={[Function]}
  contentContainerStyle={
    Object {
      "flexGrow": 1,
      "padding": 16,
    }
  }
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexGrow": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "rgb(28, 28, 30)",
          }
        }
      >
        Nothing found
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
